### PR TITLE
GCI: add support for network plugin

### PR DESCRIPTION
**Pull request by andyzheng0831**
I had run e2e against a cluster with both master and nodes on GCI a couple of times. The PR auto tests will cover the hybrid cluster with just master on GCI.

cc/ @roberthbailey @fabioy @kubernetes/goog-image

### CODE REVIEW

File: cluster/gce/gci/configure.sh [hunk 1]
```diff
@@ -146,6 +146,17 @@ function install-kube-binary-config {
   else
     rm -f "${kube_bin}/kubelet"
   fi
+  if [[ "${NETWORK_PROVIDER:-}" == "kubenet" ]] || \
+     [[ "${NETWORK_PROVIDER:-}" == "cni" ]]; then
+    #TODO(andyzheng0831): We should make the cni version number as a k8s env variable.
```
Comment (andyzheng0831) [inline cluster/gce/gci/configure.sh]: @freehan I think we should find a way to simplify the code maintenance when bumping the cni version. One option is put the version number in a kube env variable, which will be passed to the network plugin manifest and here. So you will need to just update one place whenever you need to bump the version. What do you think?

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -261,6 +268,12 @@ function assemble-docker-flags {
   fi
 
   echo "DOCKER_OPTS=\"${docker_opts} ${EXTRA_DOCKER_OPTS:-}\"" > /etc/default/docker
+  # If using a network plugin, we need to explicitly restart docker daemon, because
+  # kubelet will not do it. 
+  if [[ "${use_net_plugin=}" == "true" ]]; then
```
Comment (roberthbailey) [inline cluster/gce/gci/configure-helper.sh]: should this be `"${use_net_plugin:-}"`?

File: cluster/gce/gci/configure.sh (hunk 1 above)
Comment (roberthbailey) [inline cluster/gce/gci/configure.sh]: If cni is specified then the cni version should be another kube-env field.

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (andyzheng0831) [inline cluster/gce/gci/configure-helper.sh]: Nice catch, thanks! Will correct it.

File: cluster/gce/gci/configure.sh (hunk 1 above)
Comment (andyzheng0831) [inline cluster/gce/gci/configure.sh]: I am not familiar with the cni manifest change, so I leave a TODO here, and will coordinate with the cni owner freehan@ to make the change, which will be in a separate PR.

**Comment by andyzheng0831** [comment]
@roberthbailey @fabioy the network plugin was added since 1.2 branch but I notice currently GKE does not use the network plugin. I am not sure what is GKE's plan about this network plugin feature. If not necessary, we may avoid making a trusty change for it and cherrypick in 1.2-release branch.

File: cluster/gce/gci/configure.sh (hunk 1 above)
Comment (freehan) [inline cluster/gce/gci/configure.sh]: Ideally, we always want to use the latest stable cni version in kubenet. If customer want to specify their own, they can use cni network plugin and bring their own config/binary.

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -329,6 +343,7 @@ function start-kubelet {
       flags+=" --register-schedulable=false"
       flags+=" --reconcile-cidr=false"
       flags+=" --pod-cidr=10.123.45.0/30"
+      reconcile_cidr=false
```
Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: flags += " --reconcile-cidr=false"

Comment (andyzheng0831) [inline cluster/gce/gci/configure-helper.sh]: Ah I made a mistake that I thought the default value is false. Just check it, this flag is true by default. I will update the code

**Comment by andyzheng0831** [comment]
@freehan the PR is updated, please review it. Thanks!

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -329,6 +343,7 @@ function start-kubelet {
       flags+=" --register-schedulable=false"
       flags+=" --reconcile-cidr=false"
```
Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: you may end up with duplicate flag here

File: cluster/gce/gci/configure-helper.sh [hunk 5]
```diff
@@ -341,6 +356,15 @@ function start-kubelet {
       flags+=" --hairpin-mode=${HAIRPIN_MODE}"
     fi
   fi
+  # Network plugin
+  if [[ -n "${NETWORK_PROVIDER:-}" ]]; then
+    flags+=" --network-plugin-dir=/home/kubernetes/bin"
+    flags+=" --network-plugin=${NETWORK_PROVIDER}"
+  fi
+  flags+=" --reconcile-cidr=${reconcile_cidr}"
```
Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: you may end up with duplicate flag here

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (andyzheng0831) [inline cluster/gce/gci/configure-helper.sh]: This should be deleted. Will update it shortly

**Comment by andyzheng0831** [comment]
@k8s-bot unit test this issue #26815

**Comment by freehan** [comment]
LGTM pending on tests

**Comment by andyzheng0831** [comment]
@k8s-bot test this issue #IGNORE

**Comment by freehan** [comment]
bump priority

**Comment by freehan** [comment]
just to be safe. Can we NOT run kubenet on master? Since we are very close on release cut.

**Comment by freehan** [comment]
Nevermind. just check the code once again. It should be ok.

**Comment by andyzheng0831** [comment]
@freehan FYI, if you think network plugin is not ready to deploy, please set the NETWORK_PROVIDER to false. For GCI, our target is to provide feature parity as ContainerVM, and this is the reason why I made this PR to add the network plugin in GCI config. As for GKE rollout, this pr should be OK because currently GKE does not use kubenet.

**Comment by roberthbailey** [comment]
@matchstick -- what is the plan with network plugins for 1.3?

**Comment by freehan** [comment]
@roberthbailey kubenet is on by default now.

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -341,6 +355,15 @@ function start-kubelet {
       flags+=" --hairpin-mode=${HAIRPIN_MODE}"
     fi
   fi
+  # Network plugin
+  if [[ -n "${NETWORK_PROVIDER:-}" ]]; then
+    flags+=" --network-plugin-dir=/home/kubernetes/bin"
```
Comment (bboreham) [inline cluster/gce/gci/configure-helper.sh]: Note that for CNI this parameter sets the directory where _config_ is searched, not binaries.  As far as I can tell CNI just doesn't work on GCI now.

Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: True. Will fix it along with: https://github.com/kubernetes/kubernetes/issues/28563

Comment (bboreham) [inline cluster/gce/gci/configure-helper.sh]: Oh; I actually started on a PR as this is a total blocker for me.

Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: Go ahead then. I have not started. Please also refer to the discussions in #28563.
